include/vcpkg/base/downloads.h

```diff
@@ -52,6 +52,22 @@ namespace vcpkg::Downloads
         explicit DownloadManager(const DownloadManagerConfig& config) : m_config(config) { }
         explicit DownloadManager(DownloadManagerConfig&& config) : m_config(std::move(config)) { }
 
+        void download_file_without_hash_check(Filesystem& fs,
```

Comment:
This a classic "do_a_and_b" function, is there a reason it isn't just "do_a" followed by "do_b" ?

Reply:
I see, because the rename has to happen after the hash check. Proposal pseudocode:

```
struct SkipShaT {};
constexpr SkipShaT skip_sha;

struct ShaCheckSetting {
    ShaCheckSetting(StringView sv) : hash(sv.to_string()) {
        Checks::check_exit(VCPKG_LINE_INFO, is_sha512(sv), "%s was not a valid SHA512 hash", sv);
    }

    ShaCheckSetting(SkipShaT) : hash() {}
    // download_file calls this then renames itself
    bool check(Filesystem& fs, StringView sanitized_url, StringView download_path_part_path, string& errors) {
        if (!hash.empty())
        {
            auto maybe_error = try_verify_downloaded_file_hash(fs, sanitized_url, download_path_part_path, hash);
            if (auto err = maybe_error.get())
            {
                Strings::append(errors, *err);
                return false;
            }
        }

        return true;
    }
private:
    std::string hash;
}

void download_file(Filesystem& fs,
                    const std::string& url,
                    const path& download_path,
                    ShaCheckSetting hash) const
{
    this->download_file(fs, url, {}, download_path, hash);
}
```